lms/djangoapps/courseware/model_data.py

```diff
@@ -34,12 +35,473 @@ class InvalidWriteError(Exception):
+            field_object = self._cache.get(cache_key)
+
+            try:
+                serialized_value = json.dumps(value)
+                # It is safe to force an insert or an update, because
+                # a) we should have retrieved the object as part of the
+                #    prefetch step, so if it isn't in our cache, it doesn't exist yet.
+                # b) no other code should be modifying these models out of band of
+                #    this cache.
+                if field_object is None:
+                    field_object = self._create_object(kvs_key, serialized_value)
+                    field_object.save(force_insert=True)
+                    self._cache[cache_key] = field_object
+                else:
+                    field_object.value = serialized_value
+                    field_object.save(force_update=True)
+
+            except DatabaseError:
+                log.exception("Saving field %r failed", kvs_key.field_name)
+                raise KeyValueMultiSaveError(saved_fields)
+
+            finally:
+                saved_fields.append(kvs_key.field_name)
+
+    @contract(kvs_key=DjangoKeyValueStore.Key)
+    def delete(self, kvs_key):
+        """
+        Delete the value specified by `kvs_key`.
+
+        Arguments:
+            kvs_key (`DjangoKeyValueStore.Key`): The field value to delete
+
+        Raises: KeyError if key isn't found in the cache
+        """
+
+        field_object = self._cache.get(self._cache_key_for_kvs_key(kvs_key))
+        if field_object is None:
+            raise KeyError(kvs_key.field_name)
+
+        field_object.delete()
+
+    @contract(kvs_key=DjangoKeyValueStore.Key, returns=bool)
+    def has(self, kvs_key):
+        """
+        Return whether the specified `kvs_key` is set.
+
+        Arguments:
+            kvs_key (`DjangoKeyValueStore.Key`): The field value to delete
+
+        Returns: bool
+        """
+        return self._cache_key_for_kvs_key(kvs_key) in self._cache
+
+
+    @contract(kvs_key=DjangoKeyValueStore.Key, returns="datetime|None")
+    def last_modified(self, kvs_key):
+        """
+        Return when the supplied field was changed.
+
+        Arguments:
+            kvs_key (`DjangoKeyValueStore.Key`): The field value to delete
+
+        Returns: datetime if there was a modified date, or None otherwise
+        """
+        field_object = self._cache.get(self._cache_key_for_kvs_key(kvs_key))
+
+        if field_object is None:
+            return None
+        else:
+            return field_object.modified
+
+    def __len__(self):
+        return len(self._cache)
+
+
+class UserStateCache(object):
+    def __init__(self, user, course_id):
+        self._cache = defaultdict(dict)
+        self.course_id = course_id
+        self.user = user
+        self._client = DjangoXBlockUserStateClient(self.user)
+
+    def cache_fields(self, fields, descriptors, aside_types):
+        query = self._client.get_many(
```

Comment:
Nit: This variable reads as being the query itself, when it's actually the query results. `results`?